db/python/layers/project_insights.py

```diff
@@ -0,0 +1,667 @@
+        """
+        Filter returned records for a specific project and sequencing type and
+        return the value of a specific field
+        """
+        if not rows:
+            return 0
+        for row in rows:
+            if row['project'] == project and row['sequencing_type'] == sequencing_type:
+                return row[field]
+        return 0
+
+    @staticmethod
+    def get_val_for_project_and_sequencing_type_and_sg_id(
+        project, sequencing_type, sg_id, field, rows
+    ):
+        """
+        Filter returned records for a specific project, sequencing type and
+        sequencing group id and return the value of a specific field
+        """
+        if not rows:
+            return 0
+        for row in rows:
+            if (
+                row['project'] == project
+                and row['sequencing_type'] == sequencing_type
+                and row['sequencing_group_id'] == sg_id
+            ):
+                return row[field]
+        return 0
+
+    @staticmethod
+    def get_val_for_project_and_sequencing_type_and_stage(
+        project, sequencing_type, stage, field, rows
+    ):
+        """
+        Filter returned records for a specific project, sequencing type and
+        stage and return the value of a specific field
+        """
+        if not rows:
+            return 0
+        for row in rows:
+            if (
+                row['project'] == project
+                and row['sequencing_type'] == sequencing_type
+                and row['stage'] == stage
+            ):
+                return row[field]
+        return 0
+
+    async def get_project_insights_stats(
+        self, projects: list[int], sequencing_types: list[str]
+    ):
+        """Combines the results of the above insights stats queries into a response"""
+
+        ptable = ProjectPermissionsTable(self._connection)
+        await ptable.check_access_to_project_ids(
+            user=self.author, project_ids=projects, readonly=True
+        )
+
+        (
+            families,
+            participants,
+            samples,
+            sequencing_groups,
+            crams,
+            latest_annotate_dataset,
+            latest_es_indexes,
+        ) = await asyncio.gather(
+            self._stats_families_query(projects, sequencing_types),
+            self._stats_participants_query(projects, sequencing_types),
+            self._stats_samples_query(projects, sequencing_types),
+            self._stats_sequencing_groups_query(
+                projects, sequencing_types
+            ),
+            self._stats_crams_query(projects, sequencing_types),
+            self._stats_annotate_dataset_query(
+                projects, sequencing_types
+            ),
+            self._stats_es_indexes_query(projects, sequencing_types,),
+        )
+
+        response = []
+        for pid in projects:
+            project = await ptable.get_and_check_access_to_project_for_id(self.author, pid, readonly=True)
```

Comment:
A call to check access to multiple projects (save this async function being in a loop)